contracts/governance/proposals.cairo

```diff
@@ -0,0 +1,287 @@
+    let (curr_block_number) = get_block_number();
+    with_attr error_message("voting already concluded"){
+        // yes truly no assert_lt in Cairo 0.10.
+        let block_diff = end_block_number - curr_block_number;
+        assert_not_zero(block_diff);
+        assert_nn(block_diff);
+    }
+
+    // Cast vote
+    proposal_voted_by.write(prop_id, caller_addr, opinion);
+    if(opinion == -1){
+        let (curr_votes) = proposal_total_nay.read(prop_id);
+        let new_votes = curr_votes + caller_balance.low;
+        assert_nn(new_votes);
+        proposal_total_nay.write(prop_id, new_votes);
+    }else{
+        let (curr_votes) = proposal_total_yay.read(prop_id);
+        let new_votes = curr_votes + caller_balance.low;
+        assert_nn(new_votes);
+        proposal_total_yay.write(prop_id, new_votes);
+    }
+    return ();
+}
+
+
+@view 
+func get_proposal_details{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    prop_id: felt
+) -> (res: PropDetails) {
+    let (res) = proposal_details.read(prop_id);
+    return (res=res);
+}
+
+
+// @notice Returns proposal status – passed = 1, rejected = -1, voting = 0.
+// rejected due to not enough votes (didn't meet quorum) = -1
+// @dev fails if proposal doesn't exist
+@view
+func get_proposal_status{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    prop_id: felt
+) -> (res: felt) {
+    let (end_block_number) = proposal_vote_ends.read(prop_id);
+    let (curr_block_number) = get_block_number();
+    let block_cmp = is_le(curr_block_number, end_block_number);
+    if (block_cmp == 1){
+        return (res=0);
+    }
+    
+    let (gov_token_addr) = governance_token_address.read();
+    let (nay_tally) = proposal_total_yay.read(prop_id);
+    let (yay_tally) = proposal_total_nay.read(prop_id);
+    let total_tally = yay_tally + nay_tally;
+    let total_tally_uint256 = intToUint256(total_tally);
+    
+
+    with_attr error_message("unable to check quorum"){
+        let share = Uint256(low = QUORUM, high = 0);
+        let (tally_multiplied, carry) = uint256_mul(total_tally_uint256, share);
+        assert carry.low = 0;
+        assert carry.high = 0;
+        let (total_eligible_votes) = IERC20.totalSupply(contract_address=gov_token_addr);
+    }
+
+    let (cmp_res) = uint256_lt(total_eligible_votes, tally_multiplied);
+    if (cmp_res == 0) {
+        return (res=-1); // didn't meet quorum
+    }
+
+    let yay_or_nay = yay_tally - nay_tally;
+    if (yay_or_nay == 0){
+        return (res=-1); // yay_tally = nay_tally
+    }
+    
+    let nn = is_nn(yay_or_nay);
+    if (nn == 1) {
+        return (res=1); // yay_tally > nay_tally
+    }else{
+        return (res=-1); // yay_tally < nay_tally
+    }
+}
+
+
+// @notice Investors don't hold tokens as of launch. They can vote only from whitelisted addresses.
+// When investor votes, the real voting power of all investors is equivalent to totalSupply - balanceOf(TEAM_MULTISIG_ADDR).
```

Comment:
`totalSupply - balanceOf(TEAM_MULTISIG_ADDR)` looks like minus... but I guess it should be `investors is equivalent to balanceOf(TEAM_MULTISIG_ADDR).`